exchange/bidder_test.go

```diff
@@ -3209,3 +3213,86 @@ func TestUpdateBidderTmax(t *testing.T) {
+	deadline := time.Date(2023, 5, 30, 1, 0, 0, int(requestTmaxNS), time.UTC)
+	ctx := &mockBidderTmaxCtx{startTime: startTime, deadline: deadline, now: now, ok: true}
+
+	tests := []struct {
+		description     string
+		ctx             bidderTmaxContext
+		requestTmax     int64
+		tmaxAdjustments config.TmaxAdjustments
+		expected        bool
+	}{
+		{
+			description:     "bidder-response-duration-min-not-set",
+			ctx:             ctx,
+			requestTmax:     requestTmaxMS,
+			tmaxAdjustments: config.TmaxAdjustments{Enabled: true, BidderResponseDurationMin: 0},
+			expected:        false,
+		},
+		{
+			description:     "network-latency-buffer-not-set",
+			ctx:             ctx,
+			requestTmax:     requestTmaxMS,
+			tmaxAdjustments: config.TmaxAdjustments{Enabled: true, BidderResponseDurationMin: 10, BidderNetworkLatencyBuffer: 0},
+			expected:        false,
+		},
+		{
+			description:     "response-preparation-duration-not-set",
+			ctx:             ctx,
+			requestTmax:     requestTmaxMS,
+			tmaxAdjustments: config.TmaxAdjustments{Enabled: true, BidderResponseDurationMin: 10, BidderNetworkLatencyBuffer: 20, PBSResponsePreparationDuration: 0},
+			expected:        false,
+		},
+		{
+			description:     "remaing-duration-greater-than-bidder-response-min",
+			ctx:             ctx,
+			requestTmax:     requestTmaxMS,
+			tmaxAdjustments: config.TmaxAdjustments{Enabled: true, PBSResponsePreparationDuration: 50, BidderNetworkLatencyBuffer: 10, BidderResponseDurationMin: 40},
+			expected:        false,
+		},
+		{
+			description:     "remaing-duration-less-than-bidder-response-min",
+			ctx:             ctx,
+			requestTmax:     requestTmaxMS,
+			tmaxAdjustments: config.TmaxAdjustments{Enabled: true, PBSResponsePreparationDuration: 100, BidderNetworkLatencyBuffer: 10, BidderResponseDurationMin: 500},
+			expected:        true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) {
+			assert.Equal(t, test.expected, hasShorterDurationThanTmax(test.ctx, test.tmaxAdjustments))
+		})
+	}
+}
+
+func TestBidderTmaxTimeout(t *testing.T) {
+	d1 := time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)
+	d2 := time.Date(2023, 1, 1, 0, 0, 0, int(20*time.Millisecond), time.UTC)
+
+	ctx, cancelFunc := context.WithDeadline(context.Background(), d1)
+	defer cancelFunc()
+
+	bidRequest := adapters.RequestData{
+		Method: "POST",
+		Uri:    "test.url.com",
+		Body:   []byte(`{"id":"this-id","app":{"publisher":{"id":"pub-id"}}}`),
+	}
+
+	bidderAdapter := bidderAdapter{
+		me: &metricsConfig.NilMetricsEngine{},
+	}
+	logger := func(msg string, args ...interface{}) {}
+
+	tmaxAdjustments := &config.TmaxAdjustments{Enabled: true, PBSResponsePreparationDuration: 100, BidderNetworkLatencyBuffer: 10, BidderResponseDurationMin: 5000}
+
+	httpCallInfo := bidderAdapter.doRequestImpl(ctx, &bidRequest, logger, d2, tmaxAdjustments)
+	assert.Equal(t, errTmaxTimeout, httpCallInfo.err)
```

Comment:
I think you should add a test case for happy path where if condition that you just added passes. thoughts?

Reply:
Also since you have added the conditions in if statement, I think you should also test for each of those conditions like config.TmaxAdjustments = nil etc.

Reply:
https://github.com/prebid/prebid-server/pull/2848/commits/0ac67877a0da81f9620f3ccd1922aa9667680f89 adds above mentioned test case